src/Microsoft.AspNet.Razor.Runtime/TagHelpers/ReadOnlyTagHelperAttributes.cs

```diff
@@ -0,0 +1,119 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.Framework.Internal;
+
+namespace Microsoft.AspNet.Razor.Runtime.TagHelpers
+{
+    public class ReadOnlyTagHelperAttributes<TAttributeType> : IReadOnlyList<TAttributeType>
+        where TAttributeType : IReadOnlyTagHelperAttribute
+    {
+        protected readonly List<TAttributeType> _attributes;
+
+        protected ReadOnlyTagHelperAttributes()
+        {
+            _attributes = new List<TAttributeType>();
+        }
+
+        public ReadOnlyTagHelperAttributes([NotNull] IEnumerable<TAttributeType> attributes)
+        {
+            _attributes = new List<TAttributeType>(attributes);
+        }
+
+        public TAttributeType this[int index]
+        {
+            get
+            {
+                return _attributes[index];
+            }
+        }
+
+        public TAttributeType this[[NotNull] string key]
```

Comment:
thinking we may want to go further in @Eilon's "really not a dictionary" direction and use leave this out.  consumers have `TryGetValue()` when the want the one important `TagHelperAttribute`.

similar objection to setter for `string`-based itemizer in `TagHelperAttributes`.  suggest a `ReplaceAll(TagHelperAttribute)` method instead.